File: apps/dsa-web/src/api/price_monitor.ts
import apiClient from './index';

// ============================================
// 类型定义
// ============================================

export type PriceMonitorStatus = {
  running: boolean;
  check_interval_seconds: number;
  watching_count: number;
  triggered_cache_count: number;
};

export type WatchingStock = {
  id: number;
  code: string;
  name?: string | null;
  sentiment_score: number;
  analysis_date: string;
  ideal_buy: number;
  secondary_buy?: number | null;
  stop_loss?: number | null;
  take_profit?: number | null;
  analysis_history_id?: number | null;
  is_active: boolean;
  triggered: boolean;
  triggered_at?: string | null;
  trigger_price?: number | null;
  note?: string | null;
  created_at: string;
  updated_at: string;
};

export type PriceAlert = {
  id: number;
  watching_stock_id: number;
  code: string;
  name?: string | null;
  alert_type: 'ideal_buy' | 'secondary_buy' | 'stop_loss' | 'take_profit';
  target_price: number;
  trigger_price: number;
  change_pct?: number | null;
  analysis_triggered: boolean;
  analysis_query_id?: string | null;
  notification_sent: boolean;
  notification_sent_at?: string | null;
  created_at: string;
};

export type AddWatchingStockRequest = {
  code: string;
  name?: string;
  sentiment_score: number;
  ideal_buy: number;
  secondary_buy?: number;
  stop_loss?: number;
  take_profit?: number;
  analysis_history_id?: number;
};

export type PriceAlertListResponse = {
  items: PriceAlert[];
  total: number;
  page: number;
  limit: number;
};

export type WatchingStockListResponse = {
  items: WatchingStock[];
  total: number;
  page: number;
  limit: number;
};

export type MonitorOperationResponse = {
  success: boolean;
  message: string;
};

// ============================================
// API 方法
// ============================================

export const priceMonitorApi = {
  /** 获取服务状态 */
  async getStatus(): Promise<PriceMonitorStatus> {
    const response = await apiClient.get('/api/v1/price-monitor/status');
    return response.data as PriceMonitorStatus;
  },

  /** 启动服务 */
  async start(): Promise<MonitorOperationResponse> {
    const response = await apiClient.post('/api/v1/price-monitor/start');
    return response.data as MonitorOperationResponse;
  },

  /** 停止服务 */
  async stop(): Promise<MonitorOperationResponse> {
    const response = await apiClient.post('/api/v1/price-monitor/stop');
    return response.data as MonitorOperationResponse;
  },

  /** 刷新盯盘列表 */
  async refreshList(): Promise<MonitorOperationResponse> {
    const response = await apiClient.post('/api/v1/price-monitor/refresh?min_score=85&days=2');
    return response.data as MonitorOperationResponse;
  },

  /** 获取盯盘列表 */
  async getWatchingList(min_score?: number): Promise<WatchingStock[]> {
    const params = min_score !== undefined ? { min_score } : {};
    const response = await apiClient.get('/api/v1/price-monitor/watching', { params });
    const data = response.data as WatchingStockListResponse;
    return data.items || [];
  },

  /** 添加盯盘股票 */
  async addWatchingStock(data: AddWatchingStockRequest): Promise<WatchingStock> {
    const response = await apiClient.post('/api/v1/price-monitor/watching', data);
    return response.data as WatchingStock;
  },

  /** 删除盯盘股票 */
  async deleteWatchingStock(id: number): Promise<MonitorOperationResponse> {
    const response = await apiClient.delete(`/api/v1/price-monitor/watching/${id}`);
    return response.data as MonitorOperationResponse;
  },

  /** 删除所有盯盘股票 */
  async deleteAllWatchingStocks(): Promise<MonitorOperationResponse> {
    const response = await apiClient.delete('/api/v1/price-monitor/watching');
    return response.data as MonitorOperationResponse;
  },

  /** 获取提醒历史 */
  async getAlerts(
    params: {
      page?: number;
      page_size?: number;
      code?: string;
      alert_type?: string;
    } = {}
  ): Promise<PriceAlertListResponse> {
    const queryParams: Record<string, any> = {};
    if (params.page) queryParams.page = params.page;
    if (params.page_size) queryParams.limit = params.page_size;
    if (params.code) queryParams.code = params.code;
    if (params.alert_type) queryParams.alert_type = params.alert_type;

    const response = await apiClient.get('/api/v1/price-monitor/alerts', { params: queryParams });
    return response.data as PriceAlertListResponse;
  },
};

// ============================================
// 工具函数
// ============================================

export function getAlertTypeLabel(type: string): string {
  const labels: Record<string, string> = {
    ideal_buy: '理想买入',
    secondary_buy: '次级买入',
    stop_loss: '止损',
    take_profit: '止盈',
  };
  return labels[type] || type;
}

export function getAlertTypeVariant(type: string): 'success' | 'warning' | 'danger' | 'info' | 'default' {
  const variants: Record<string, 'success' | 'warning' | 'danger' | 'info' | 'default'> = {
    ideal_buy: 'success',
    secondary_buy: 'info',
    stop_loss: 'danger',
    take_profit: 'warning',
  };
  return variants[type] || 'default';
}

File: api/v1/schemas/price_monitor.py
# -*- coding: utf-8 -*-
"""
===================================
价格盯盘 API Schema
===================================
"""

from typing import Optional, List, Any
from datetime import date, datetime

from pydantic import BaseModel, Field


class WatchingStockBase(BaseModel):
    """盯盘股票基础模型"""
    code: str = Field(..., description="股票代码")
    name: Optional[str] = Field(None, description="股票名称")
    sentiment_score: int = Field(..., ge=0, le=100, description="评分")
    ideal_buy: float = Field(..., gt=0, description="理想买入价格")
    secondary_buy: Optional[float] = Field(None, description="次级买入价格")
    stop_loss: Optional[float] = Field(None, description="止损价格")
    take_profit: Optional[float] = Field(None, description="止盈价格")
    analysis_history_id: Optional[int] = Field(None, description="关联的分析历史ID")
    note: Optional[str] = Field(None, description="备注")


class WatchingStockCreate(WatchingStockBase):
    """创建盯盘股票请求"""
    analysis_date: Optional[date] = Field(None, description="分析日期")


class WatchingStockUpdate(BaseModel):
    """更新盯盘股票请求"""
    name: Optional[str] = None
    sentiment_score: Optional[int] = Field(None, ge=0, le=100)
    ideal_buy: Optional[float] = Field(None, gt=0)
    secondary_buy: Optional[float] = None
    stop_loss: Optional[float] = None
    take_profit: Optional[float] = None
    is_active: Optional[bool] = None
    note: Optional[str] = None


class WatchingStockResponse(BaseModel):
    """盯盘股票响应"""
    id: int
    code: str
    name: Optional[str]
    sentiment_score: int
    analysis_date: Optional[str]
    ideal_buy: Optional[float]
    secondary_buy: Optional[float]
    stop_loss: Optional[float]
    take_profit: Optional[float]
    analysis_history_id: Optional[int]
    is_active: bool
    triggered: bool
    triggered_at: Optional[str]
    trigger_price: Optional[float]
    note: Optional[str]
    created_at: Optional[str]
    updated_at: Optional[str]

    @classmethod
    def from_dict(cls, data: dict) -> "WatchingStockResponse":
        return cls(**data)


class WatchingStockListResponse(BaseModel):
    """盯盘股票列表响应"""
    total: int
    page: int
    limit: int
    items: List[WatchingStockResponse]


class PriceAlertResponse(BaseModel):
    """价格提醒响应"""
    id: int
    watching_stock_id: int
    code: str
    name: Optional[str]
    alert_type: str
    target_price: float
    trigger_price: float
    change_pct: Optional[float]
    analysis_triggered: bool
    analysis_query_id: Optional[str]
    notification_sent: bool
    notification_sent_at: Optional[str]
    created_at: Optional[str]


class PriceAlertListResponse(BaseModel):
    """价格提醒列表响应"""
    total: int
    page: int
    limit: int
    items: List[PriceAlertResponse]


class MonitorStatusResponse(BaseModel):
    """盯盘服务状态响应"""
    running: bool
    check_interval_seconds: int
    watching_count: int
    triggered_cache_count: int


class MonitorOperationResponse(BaseModel):
    """盯盘操作响应"""
    success: bool
    message: str


class RefreshRequest(BaseModel):
    """刷新盯盘列表请求"""
    min_score: Optional[int] = Field(None, ge=0, le=100, description="最低评分")
    days: Optional[int] = Field(None, ge=1, le=30, description="回溯天数")
